src/org/purefn/river/batch.clj

```diff
@@ -0,0 +1,137 @@
+(ns org.purefn.river.batch
+  "A consumer suitable for batch reading operations from Kafka."
+  (:require [clojure.pprint :as pprint]
+            [clojure.spec.alpha :as s]
+            [com.stuartsierra.component :as component]
+            ;; temp, remove
+            [org.purefn.river.serdes.nippy :as serdes]
+            ;;----
+            [taoensso.timbre :as log])
+  (:import [org.apache.kafka.clients.consumer KafkaConsumer]
+           [org.apache.kafka.common TopicPartition]
+           [org.apache.kafka.common.errors WakeupException]
+           [org.purefn.river Message]))
+
+(defn kafka-consumer
+  [group-id servers]
+  (KafkaConsumer. {"auto.offset.reset" "earliest"
+                   "bootstrap.servers" servers
+                   "enable.auto.commit" false
+                   "group.id" group-id
+                   "client.id" group-id}
+                  (serdes/nippy-deserializer)
+                  (serdes/nippy-deserializer)))
+
+(defn create-consumers
+  [{:keys [::topics ::group-id ::threads ::bootstrap-servers] :as config}]
+  (log/info config)
+  (->> (range threads)
+       (map (fn [_]
+              (log/info "Creating consumer" {:group-id group-id
+                                             :topics topics})
+                (doto (kafka-consumer group-id bootstrap-servers)
+                  (.subscribe topics))))))
+
+(defn process
+  [^KafkaConsumer consumer
+   ^clojure.lang.Atom closing
+   {:keys [::timeout] :as config}
+   dependencies
+   process-fn]
+  (let [commit #(.commitAsync consumer)
+        pfn (if (= 4 (->> (meta process-fn)
+                          (:arglists)
+                          (map count)
+                          (reduce max)))
+              (partial process-fn dependencies)
+              process-fn)]
+    (try
+      (loop [state {}]
```

Comment:
Might want to split out the idea of processing from the initialization here. I can foresee situations where you'd want to be able to do things like capture the state and save snapshots, or access it from another component to serve up stats or analytics or whatever. The state is kind of trapped inside the closure here.

Reply:
I agree that this would be useful.  I think I'll punt on it for now until the right way to do that becomes more clear, as dictated by the usage of stats/analytics.